intelligence/ts/src/flowerintelligence.ts

```diff
@@ -191,64 +179,35 @@ export class FlowerIntelligence {
       return this.getOrCreateRemoteEngine(canonicalModelId);
     }
 
-    if (this.canRunLocally(canonicalModelId)) {
-      return this.chooseLocalEngine(canonicalModelId);
+    const localEngineResult = await this.chooseLocalEngine(canonicalModelId);
+    if (localEngineResult.ok) {
+      return localEngineResult;
     }
 
     return this.getOrCreateRemoteEngine(canonicalModelId);
   }
 
-  private chooseLocalEngine(modelId: string): Result<[Engine, string]> {
-    const localProvider = Object.keys(this.#localEngines).find(
-      (provider) =>
-        this.#localEngines[provider].length > 0 && // check if the local engine exists
-        provider in this.#availableModels[modelId]
-    );
-
-    if (!localProvider) {
-      return {
-        ok: false,
-        failure: {
-          description: `The model "${modelId}" is not available for local inference.`,
-          code: FailureCode.NoLocalProviderError,
-        },
-      };
+  private async chooseLocalEngine(modelId: string): Promise<Result<[Engine, string]>> {
+    const compatibleEngines = [];
+    for (const engine of this.#availableLocalEngines) {
+      const supportedResult = await engine.isSupported(modelId);
+      if (supportedResult.ok) {
+        compatibleEngines.push([engine, supportedResult.value] as [Engine, string]);
+      }
     }
```

Comment:
How about using either `Promise.all` to run in parallel?

```suggestion
  const compatibleEngines = await Promise.all(
    this.#availableLocalEngines.map(async (engine) => {
      const supportedResult = await engine.isSupported(modelId);
      return supportedResult.ok ? [engine, supportedResult.value] : null;
    })
  );

  // Filter out null values
  const filteredCompatibleEngines = compatibleEngines.filter(Boolean);
```